### ISSUE TITLE
Metro config can't resolve the correct `transformer.asyncRequireModulePath` path

### ISSUE THREAD

**Issue by yusufyildirim** [expo/expo#25498]
### Minimal reproducible example

can add later if necessary

### Summary

I faced with an issue while transitioning our monorepo to PNPM. Started to experience this vague bundling error:

<img width="969" alt="Screenshot 2023-11-21 at 13 11 29" src="https://github.com/expo/expo/assets/22980987/e9f88047-4012-4613-97bf-0fd37d34eefc">

Realised that everything works fine if I use official `@react-native/metro-config` for the metro config instead of `expo/metro-config`.  Upon further investigation I found that Expo's metro config is not resolving `transformer.asyncRequireModulePath` field correctly. It resolves to `metro-runtime/src/modules/asyncRequire` while the official `@react-native/metro-config` package resolves it to the correct location which is something like `/Users/yusufyildirim/development/xite-pmt-mobile/development/node_modules/.pnpm/metro-runtime@0.76.8/node_modules/metro-runtime/src/modules/asyncRequire.js`.

I dug a bit deeper and found that the field is resolved like: https://github.com/facebook/react-native/blob/main/packages/metro-config/index.js#L72 in the official config, they're overriding the base `metro` config while Expo's not doing that in it's own config: https://github.com/expo/expo/blob/main/packages/%40expo/metro-config/src/ExpoMetroConfig.ts#L173. 

So, the value is directly coming from the base config, which is here: https://github.com/facebook/metro/blob/main/packages/metro-config/src/defaults/index.js#L92C10-L92C10

The solution is somewhat simple, either Expo should do what's official react-native config is doing and override it on it's own config package or solve the problem directly in the metro itself.

For those who have been experiencing this, there's an easy workaround:
<img width="761" alt="Screenshot 2023-11-21 at 13 28 39" src="https://github.com/expo/expo/assets/22980987/67319bd1-6572-4070-b6e1-0b5a69470446">
I've been basically getting the correctly resolved field from official rn config package and overriding the Expo's config in my own `metro.config.js`, before exporting it.

### Environment

expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 16.20.2 - ~/.asdf/installs/nodejs/16.20.2/bin/node
      Yarn: 1.22.19 - /opt/homebrew/bin/yarn
      npm: 8.19.4 - ~/.asdf/plugins/nodejs/shims/npm
      Watchman: 2023.10.23.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /Users/yusufyildirim/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: 2022.2 AI-222.4459.24.2221.9862592
      Xcode: 14.2/14C18 - /usr/bin/xcodebuild
    Expo Workflow: managed (not really)

**Comment by byCedric** [comment]
Hi @yusufyildirim, thanks for reporting this issue! You are right, the `config.transformer.asyncRequireModulePath` should be the full path to the module, not just `metro-runtime/src/modules/asyncRequire`.

Unfortunately, this has been a bug in Metro, not Expo. Meta has put a lot of effort into React Native 0.73 (RC), and this release includes the refactored packages (from `@react-native/*`). The bugfix you [linked](https://github.com/facebook/react-native/blob/7625a502960e6b107e77542ff0d6f40fbf957322/packages/metro-config/index.js#L72) to is included in `@react-native/metro-config`, [but not in `metro-config`](https://cdn.jsdelivr.net/npm/metro-config@0.76.8/src/defaults/index.js) (_see line 84_).

Expo SDK 49 is still the latest version, which uses `metro-config@0.76.8` - even though newer patch versions are available. It's likely an issue with the whole dependency chain, which is fixed in RN 0.73.

<img width="864" alt="image" src="https://github.com/expo/expo/assets/1203991/eed125b2-b8c8-4e43-b23f-50c32c295d83">

So TL;DR; - Expo is using this specific value from the default Metro config, but RN <=0.72 has quite a few issues related to isolated modules (pnpms default behavior). Both Meta, Expo, and a few others have put some effort into fixing it.

It's best to use your workaround for now when you run into these issues. This will be fully fixed in SDK 50.

Hope this helps!
